neps/nep-0539.md

```diff
@@ -0,0 +1,307 @@
+system.
+
+First attempts to limit the memory consumption have been added without protocol
+changes. This is known as "local congestion control" and can be summarized in
+two rules:
+
+- Limit the transaction pool to 100 MB.
+  https://github.com/near/nearcore/pull/9172
+- Once we have accumulated more than 20k delayed receipts in a shard,
+  chunk-producers for that shard stop accepting more transactions.
+  https://github.com/near/nearcore/pull/9222
+
+But this does not put any limits on what shards will accept for other shards.
+For example, when a particular contract is popular, the contract's shard will
+eventually stop accepting new transactions but all other shards keep accepting
+more and more transactions that produce receipts for the popular contract.
+Therefore the number of delayed receipts keeps growing indefinitely.
+
+Cross-shard congestion control addresses this issue by stopping new transactions
+at the source and delaying receipt forwarding when the receiving shard has
+reached its memory limit.
+
+## Specification
+
+The proposal adds fields to chunks headers, it adds a trie columns, it changes
+the rules to select transactions, and it changes the chunk execution flow. The
+next four sections specifies each of these changes in more details.
+
+### Chunk header changes
+
+We change the chunk header to include congestion information.
+Specifically, we include two indicators.
+
+```rust
+ShardChunkHeaderInnerV3 {
+  // ...all field from ShardChunkHeaderInnerV2
+  incoming_congestion: u16
+  general_congestion: u16
+}
+```
+
+This adds 4 bytes to the chunk header, increasing it from 374 bytes to 378
+bytes in borsh representation. (Assuming no validator proposals included.)
+This in turn increases the block size by 4 bytes per shard.
+
+While the numbers are stored as `u16`,for the remainder of the NEP
+specification, `incoming_congestion` and `general_congestion` values are
+interpreted as numbers between 0 and 1, dividing the stored `u16` by `u16::MAX`
+(65535).
+
+### Changes to transaction selection
+
+Today, transactions are taken from the chunk producer's pool until `tx_limit` is
+reached, where `tx_limit` is computed as follow.
+
+```python
+# old
+tx_limit = 500 Tgas if len(delayed_receipts) < 20_000 else 0
+```
+
+We replace the formula for the transaction limit to depend on the
+`general_congestion` computed in the previous shard variable (between 0 and 1):
+
+```python
+# new
+MIN_GAS = 5 Tgas
+MAX_GAS = 500 Tgas
+tx_limit = mix(MAX_GAS, MIN_GAS, general_congestion)
+```
+
+*TODO: fine-tune `MIN_GAS` and `MAX_GAS`*
+
+In the pseudo code above, we borrow the [`mix`](https://docs.gl/sl4/mix)
+function from GLSL for linear interpolation.
+
+> `mix(x, y, a)`
+> 
+> `mix` performs a linear interpolation between x and y using a to weight between
+> them. The return value is computed as $x \times (1 - a) + y \times a$. 
+
+
+Plus, we add the additional rule to reject all transactions to shard with an
```

Comment:
Kind of yes, in the proposal we are only rejecting it at the chunk producer level. Everything else is not a protocol change and will not make it into this NEP. So we don't really define what should be treated as accepted or not accepcted.

Your described ideal world can certainly be implemented with the proposed spec here. If we take the vocabulary from [issue #6837](https://github.com/near/nearcore/issues/6837), anything except wait_until=None should not return success before a transaction is actually in a block. And the rejection filter happens before inclusion in a block.

In the most naive solution, we wait until the chunk producer rejects it and don't tell the RPC node waiting for it. This would be bad and the transaction just keeps waiting until some timeout.

Middle-ground: We signal back to the RPC node that the tx was rejected. This can be tricky since we forward it to several chunk producers and maybe some of them accept while others reject, if congestion is just building up or just being resolved. But we can just wait for all of them to answer before telling the user that it was rejected.

Dream scenario: RPC nodes run the rejection filter before even forwarding and reject it instantly. This also reduced load on the network to chunk producers.

Reply:
sounds good.  Agreed that such discussions are outside the scope of the NEP and should definitely be possible with this NEP.